packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts

```diff
@@ -135,59 +116,39 @@ export class VirtualService extends cdk.Resource implements IVirtualService {
 
   public readonly clientPolicy?: ClientPolicy;
 
-  private readonly virtualServiceProvider?: CfnVirtualService.VirtualServiceProviderProperty;
-
   constructor(scope: Construct, id: string, props: VirtualServiceProps) {
     super(scope, id, {
       physicalName: props.virtualServiceName || cdk.Lazy.string({ produce: () => cdk.Names.uniqueId(this) }),
     });
 
-    if (props.virtualNode && props.virtualRouter) {
-      throw new Error('Must provide only one of virtualNode or virtualRouter for the provider');
-    }
-
-    this.mesh = props.mesh;
     this.clientPolicy = props.clientPolicy;
+    const provider = props.virtualServiceProvider?.bind(this);
+    this.mesh = provider.mesh;
 
-    // Check which provider to use node or router (or neither)
-    if (props.virtualRouter) {
-      this.virtualServiceProvider = this.addVirtualRouter(props.virtualRouter.virtualRouterName);
-    }
-    if (props.virtualNode) {
-      this.virtualServiceProvider = this.addVirtualNode(props.virtualNode.virtualNodeName);
+    if (provider.mesh != this.mesh) {
```

Comment:
This check doesn't make sense to me... the code is

```ts
    this.mesh = provider.mesh;
    if (provider.mesh != this.mesh) {
      // ...
    }
```

So the above `if` will never have its condition be true.

Reply:
Ah good call. Fixed